content/blog/2022/2022-01-25-annotation-processing.md

```diff
@@ -0,0 +1,951 @@
+
+```java
+@Company{    
+    name="ABC"
+    city="XYZ"
+}
+public class CustomAnnotatedEmployee { .. }
+```
+
+To create a Custom annotation we need to declare it with the @interface keyword as below:
+
+```java
+public @interface Company{
+}
+```
+
+To specify information about the scope of the annotation and the area it targets, such as compile-time or run-time, we need to add Meta-Annotations to the Custom Annotation. For eg., to specify that the Annotation applies to Classes only, we need to add @Target(ElementType.TYPE), which specifies that this annotation only applies to Classes, and @Retention(RetentionPolicy.RUNTIME), which specifies that this annotation only applies at run-time. We will discuss further details about Meta-Annotations once we get this basic example running. The basic annotation is:
+
+```java
+@Target(ElementType.TYPE)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface Company{
+}
+```
+
+Next, we need to add the fields to the Custom annotation. In this case, we need 'name' and 'city'. So we add it as below:
+
+```java
+@Target(ElementType.TYPE)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface Company{
+	String name() default "ABC";
+	String city() default "XYZ";
+}
+```
+
+Putting it all together, we create a CustomAnnotatedEmployee class and apply the annotation to it as below:
+
+```java
+@Company
+public class CustomAnnotatedEmployee {
+
+    private int id;
+    private String name;
+
+    public CustomAnnotatedEmployee(int id, String name) {
+        this.id = id;
+        this.name = name;
+    }
+
+    public void getEmployeeDetails(){
+
+        System.out.println("Employee Id: " + id);
+        System.out.println("Employee Name: " + name);
+    }
+}
+```
+
+Now we create a test class to test this out:
+
+```java
+import java.lang.annotation.Annotation;
+
+public class TestCustomAnnotatedEmployee {
+
+    public static void main(String[] args) {
+
+        CustomAnnotatedEmployee employee = new CustomAnnotatedEmployee(1, "John Doe");
+        employee.getEmployeeDetails();
+
+        Annotation companyAnnotation = employee.getClass().getAnnotation(Company.class);
+        Company company = (Company)companyAnnotation;
+
+        System.out.println("Company Name: " + company.name());
+        System.out.println("Company City: " + company.city());
+    }
+}
+```
+
+This would give the output below:
+
+```java
+Employee Id: 1
+Employee Name: John Doe
+Company Name: ABC
+Company City: XYZ
+```
+
+So by introspecting the annotation at run-time we can access some common information of all employees and avoid the a lot of repetition if we had to construct a lot of objects.
+
+Now, we will get into the details of Meta-Annotations, which we used in the example above:
+
+### Meta Annotations
```

Comment:
This is the only subsection of its parent section. That's a sign that the structure of the article could be better (i.e. more helpful to the reader). I would suggest pulling this section and its subsections up one level (i.e. H2 instead of H3). That would have the benefit that the TOC shows the subsections, because the TOC only goes to Level 3.